File: examples/dekita/.claude/hooks/issue-label-check.py
#!/usr/bin/env python3
"""gh issue create時に--labelオプションの指定を強制する。

Why:
    ラベルなしのIssueは分類・検索・優先度管理が困難になる。
    Issue作成時にラベルを強制することで、Issue管理の質を維持する。

What:
    - gh issue createコマンドを検出
    - --labelオプションの有無をチェック
    - ラベルがない場合、タイトル/ボディから適切なラベルを自動提案
    - ブロックし、推奨コマンドを表示

Remarks:
    - ブロック型フック
    - issue-priority-label-checkは優先度ラベル専用、本フックはラベル有無の確認

Changelog:
    - silenvx/dekita#xxx: フック追加
    - silenvx/dekita#2451: タイトル/ボディからラベル自動提案機能を追加
"""

import json
import os.path
import shlex
import sys

from lib.execution import log_hook_execution
from lib.labels import (
    extract_body_from_command,
    extract_title_from_command,
    suggest_labels_from_text,
)
from lib.results import make_approve_result, make_block_result
from lib.session import parse_hook_input

HOOK_NAME = "issue-label-check"


def _is_gh_command(token: str) -> bool:
    """Check if a token represents the gh command (bare name or full path).

    Uses os.path.basename to correctly extract the executable name,
    avoiding false positives for commands like /usr/bin/fakegh.
    """
    # Get the basename (executable name) from the token
    # This handles both "gh" and "/usr/local/bin/gh" correctly
    # and avoids false positives like "/usr/bin/fakegh"
    return os.path.basename(token) == "gh"


def _skip_env_prefixes(parts: list[str]) -> list[str]:
    """Skip VAR=value environment variable prefixes from token list."""
    cmd_start = 0
    for i, token in enumerate(parts):
        if "=" in token and not token.startswith("-"):
            cmd_start = i + 1
        else:
            break
    return parts[cmd_start:]


def is_gh_issue_create_command(command: str) -> bool:
    """
    コマンドが実際に gh issue create で始まるかチェック。

    単純な部分文字列マッチングではなく、トークン化して
    先頭のコマンドが gh issue create であることを確認する。
    これにより、コミットメッセージや引数内に「gh issue create」が
    含まれていても誤検知しない。

    環境変数プレフィックス（例: GH_TOKEN=xxx gh issue create）も
    正しく検出する。

    フルパス指定（例: /usr/local/bin/gh issue create）も
    正しく検出する。
    """
    try:
        tokens = shlex.split(command)
    except ValueError:
        # クォートが閉じていない等の場合はフォールバック
        parts = command.split()
        # 環境変数プレフィックスをスキップ
        remaining = _skip_env_prefixes(parts)
        if len(remaining) < 3:
            return False
        return _is_gh_command(remaining[0]) and remaining[1] == "issue" and remaining[2] == "create"

    # 環境変数プレフィックス（VAR=value形式）をスキップ
    remaining = _skip_env_prefixes(tokens)

    # 残りのトークンが3つ以上あり、gh issue create かチェック
    if len(remaining) < 3:
        return False
    return _is_gh_command(remaining[0]) and remaining[1] == "issue" and remaining[2] == "create"


def has_label_option(command: str) -> bool:
    """
    コマンドに --label オプションが指定されているかチェック。

    タイトルやボディ内の文字列ではなく、実際のオプションとして
    指定されているかを判定する。
    """
    try:
        tokens = shlex.split(command)
    except ValueError:
        # クォートが閉じていない等の場合はフォールバック
        return "--label" in command.split() or "-l" in command.split()

    # --label または -l がトークンとして存在するかチェック
    for i, token in enumerate(tokens):
        # --label value または --label=value
        if token == "--label" or token.startswith("--label="):
            return True
        # -l value（短縮形）
        if token == "-l" and i + 1 < len(tokens):
            return True
    return False


def main():
    """PreToolUse hook for Bash commands."""
    try:
        data = parse_hook_input()
        tool_input = data.get("tool_input", {})
        command = tool_input.get("command", "")

        # gh issue create コマンドを検出（先頭のコマンドをチェック）
        if not is_gh_issue_create_command(command):
            # 対象外のコマンドは何も出力しない
            sys.exit(0)

        # --label オプションが指定されているかチェック
        if has_label_option(command):
            # ラベルあり: 許可（出力なし）
            log_hook_execution(HOOK_NAME, "approve")
            sys.exit(0)

        # ラベルなしでIssue作成しようとしている
        # タイトル/ボディからラベルを提案
        title = extract_title_from_command(command)
        body = extract_body_from_command(command)
        suggestions = suggest_labels_from_text(title or "", body)

        reason_lines = [
            "Issue作成時に --label オプションが指定されていません。",
            "",
        ]

        if suggestions:
            # 提案がある場合はそれを表示
            suggested_labels = [label for label, _ in suggestions]
            reason_lines.append("**📝 内容から検出したラベル候補:**")
            reason_lines.append("")
            for label, description in suggestions:
                reason_lines.append(f"- `{label}`: {description}")
            reason_lines.append("")
            reason_lines.append("**推奨コマンド（優先度ラベルを追加してください）:**")
            reason_lines.append("")
            reason_lines.append("```bash")
            # 優先度P2を追加した推奨コマンドを生成
            all_labels = ",".join(suggested_labels + ["P2"])
            if title:
                # shlex.quoteで安全にエスケープ（コマンドインジェクション対策）
                escaped_title = shlex.quote(title)
                reason_lines.append(
                    f'gh issue create --title {escaped_title} --body "..." --label "{all_labels}"'
                )
            else:
                reason_lines.append(
                    f'gh issue create --title "..." --body "..." --label "{all_labels}"'
                )
            reason_lines.append("```")
            reason_lines.append("")
            reason_lines.append("**優先度の選択:**")
        else:
            # 提案がない場合は利用可能なラベル一覧を表示
            reason_lines.append("利用可能なラベルを確認してください:")
            reason_lines.append("")
            reason_lines.append("```")
            reason_lines.append("gh label list")
            reason_lines.append("```")
            reason_lines.append("")
            reason_lines.append("**主なラベル:**")
            reason_lines.append("")
            reason_lines.append("- `bug`: バグ報告")
            reason_lines.append("- `enhancement`: 新機能")
            reason_lines.append("- `documentation`: ドキュメント改善")
            reason_lines.append("")
            reason_lines.append("**優先度（必須）:**")

        reason_lines.append("")
        reason_lines.append("| 優先度 | 説明 |")
        reason_lines.append("|--------|------|")
        reason_lines.append("| P0 | Critical - 即座に対応 |")
        reason_lines.append("| P1 | High - 早急に対応 |")
        reason_lines.append("| P2 | Medium - 通常の優先度（迷ったらこれ） |")
        reason_lines.append("| P3 | Low - 時間があれば対応 |")

        reason = "\n".join(reason_lines)
        result = make_block_result(HOOK_NAME, reason)
        log_hook_execution(HOOK_NAME, "block", "label option missing")
        print(json.dumps(result))
        sys.exit(0)

    except Exception as e:
        print(f"[{HOOK_NAME}] Hook error: {e}", file=sys.stderr)
        result = make_approve_result(HOOK_NAME, f"Hook error: {e}")
        log_hook_execution(HOOK_NAME, "approve", f"Hook error: {e}")
        print(json.dumps(result))
        sys.exit(0)


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/issue_label_check.ts
#!/usr/bin/env bun
/**
 * gh issue create時に--labelオプションの指定を強制する。
 *
 * Why:
 *   ラベルなしのIssueは分類・検索・優先度管理が困難になる。
 *   Issue作成時にラベルを強制することで、Issue管理の質を維持する。
 *
 * What:
 *   - gh issue createコマンドを検出
 *   - --labelオプションの有無をチェック
 *   - ラベルがない場合、タイトル/ボディから適切なラベルを自動提案
 *   - ブロックし、推奨コマンドを表示
 *
 * Remarks:
 *   - ブロック型フック
 *   - issue-priority-label-checkは優先度ラベル専用、本フックはラベル有無の確認
 *   - Python版: issue_label_check.py
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#2451: タイトル/ボディからラベル自動提案機能を追加
 *   - silenvx/dekita#2917: TypeScript版初期実装
 *   - silenvx/dekita#3079: option_parser完全移行、gh_utilsへ共通関数抽出
 */

import { formatError } from "../lib/format_error";
import { isGhIssueCreateCommand } from "../lib/gh_utils";
import {
  extractBodyFromCommand,
  extractTitleFromCommand,
  suggestLabelsFromText,
} from "../lib/labels";
import { logHookExecution } from "../lib/logging";
import { type OptionDef, hasOption, parseOptions, tokenize } from "../lib/option_parser";
import { makeApproveResult, makeBlockResult } from "../lib/results";
import { parseHookInput } from "../lib/session";

const HOOK_NAME = "issue-label-check";

/** --label オプション定義（複数指定可能） */
const LABEL_OPTION_DEFS: OptionDef[] = [
  { long: "label", short: "l", hasValue: true, multiple: true },
];

/**
 * Check if command has --label option using option_parser.
 */
export function hasLabelOption(command: string): boolean {
  const tokens = tokenize(command);
  if (tokens.length === 0) {
    // Empty or whitespace-only command: use simple split as conservative fallback
    // Only detect = forms to avoid false positives with incomplete commands like "gh issue create -l"
    const parts = command.split(/\s+/);
    return parts.some((p) => p.startsWith("--label=") || p.startsWith("-l="));
  }
  const options = parseOptions(tokens, LABEL_OPTION_DEFS);
  return hasOption(options, "label");
}

/**
 * Escape a string for shell use (simple quoting).
 */
export function shellQuote(str: string): string {
  // If string contains single quotes, use double quotes with escaping
  if (str.includes("'")) {
    return `"${str.replace(/"/g, '\\"')}"`;
  }
  // Otherwise use single quotes
  return `'${str}'`;
}

async function main(): Promise<void> {
  let sessionId: string | undefined;
  try {
    const data = await parseHookInput();
    sessionId = data.session_id;
    const toolInput = (data.tool_input as Record<string, unknown>) || {};
    const command = (toolInput.command as string) || "";

    // Only check gh issue create commands
    if (!isGhIssueCreateCommand(command)) {
      // Not a target command, exit silently
      process.exit(0);
    }

    // Check if --label option is specified
    if (hasLabelOption(command)) {
      // Has label, approve
      await logHookExecution(HOOK_NAME, "approve", undefined, undefined, { sessionId });
      process.exit(0);
    }

    // No label specified, suggest labels from title/body
    const title = extractTitleFromCommand(command);
    const body = extractBodyFromCommand(command);
    const suggestions = suggestLabelsFromText(title || "", body);

    const reasonLines: string[] = ["Issue作成時に --label オプションが指定されていません。", ""];

    if (suggestions.length > 0) {
      const suggestedLabels = suggestions.map((s) => s.label);
      reasonLines.push("**📝 内容から検出したラベル候補:**");
      reasonLines.push("");
      for (const { label, description } of suggestions) {
        reasonLines.push(`- \`${label}\`: ${description}`);
      }
      reasonLines.push("");
      reasonLines.push("**推奨コマンド（優先度ラベルを追加してください）:**");
      reasonLines.push("");
      reasonLines.push("```bash");
      // Generate recommended command with P2 priority
      const allLabels = [...suggestedLabels, "P2"].join(",");
      if (title) {
        const escapedTitle = shellQuote(title);
        reasonLines.push(
          `gh issue create --title ${escapedTitle} --body "..." --label "${allLabels}"`,
        );
      } else {
        reasonLines.push(`gh issue create --title "..." --body "..." --label "${allLabels}"`);
      }
      reasonLines.push("```");
      reasonLines.push("");
      reasonLines.push("**優先度の選択:**");
    } else {
      // No suggestions, show available labels
      reasonLines.push("利用可能なラベルを確認してください:");
      reasonLines.push("");
      reasonLines.push("```");
      reasonLines.push("gh label list");
      reasonLines.push("```");
      reasonLines.push("");
      reasonLines.push("**主なラベル:**");
      reasonLines.push("");
      reasonLines.push("- `bug`: バグ報告");
      reasonLines.push("- `enhancement`: 新機能");
      reasonLines.push("- `documentation`: ドキュメント改善");
      reasonLines.push("");
      reasonLines.push("**優先度（必須）:**");
    }

    reasonLines.push("");
    reasonLines.push("| 優先度 | 説明 |");
    reasonLines.push("|--------|------|");
    reasonLines.push("| P0 | Critical - 即座に対応 |");
    reasonLines.push("| P1 | High - 早急に対応 |");
    reasonLines.push("| P2 | Medium - 通常の優先度（迷ったらこれ） |");
    reasonLines.push("| P3 | Low - 時間があれば対応 |");

    const reason = reasonLines.join("\n");
    const result = makeBlockResult(HOOK_NAME, reason);
    await logHookExecution(HOOK_NAME, "block", "label option missing", undefined, { sessionId });
    console.log(JSON.stringify(result));
    process.exit(0);
  } catch (error) {
    console.error(`[${HOOK_NAME}] Hook error: ${formatError(error)}`);
    const result = makeApproveResult(HOOK_NAME, `Hook error: ${formatError(error)}`);
    await logHookExecution(HOOK_NAME, "approve", `Hook error: ${formatError(error)}`, undefined, {
      sessionId,
    });
    console.log(JSON.stringify(result));
    process.exit(0);
  }
}

if (import.meta.main) {
  main();
}
